ibr/ibr-sdk/build/ibr-sdk.js

```diff
@@ -22,6 +25,164 @@ typeof define === 'function' && define.amd ? define(['exports'], factory) :
+    rootSpan.setAttribute('class', 'arrow');
+    li.appendChild(rootSpan);
+    const ul = document.createElement('ul');
+    ul.setAttribute('class', 'nested');
+    ul.setAttribute('id', ibrData.name);
+    li.appendChild(ul);
+    rootSpan.addEventListener('click', function() {
+      rootSpan.parentElement.querySelector('.nested').classList
+          .toggle('active');
+      rootSpan.classList.toggle('expanded-arrow');
+    });
+    // root structure index 0
+    extractSingleStructureData(ibrData, ibrData.name, scene, 0);
+  }
+
+  /**
+   * Generate Scene configured to display IBR data.
+   * @param {HTMLElement} parentElement parent HTML element that the
+   visualization will be append on.
+   * @return {Object} scene Scene generated to
+   */
+  function generateScene(parentElement) {
+    const scene = new THREE.Scene();
+    const camera = new THREE.PerspectiveCamera(
+        75, window.innerWidth / window.innerHeight, 0.1, 10000 );
+    const renderer = new THREE.WebGLRenderer();
+    renderer.setSize( window.innerWidth, window.innerHeight );
+    parentElement.appendChild( renderer.domElement );
+    const controls = new OrbitControls( camera, renderer.domElement );
+    controls.target.set( 0, 0.5, 0 );
+    controls.update();
+    controls.enablePan = false;
+    controls.enableDamping = true;
+    camera.position.set( 0, 0, 7000 );
+    camera.lookAt( 0, 0, 0 );
+    animate();
+
+    /**
+    * Renders the scene.
+    */
+    function animate() {
+      requestAnimationFrame( animate );
+      controls.update();
+      renderer.render( scene, camera );
+    }
+
+    return scene;
+  }
+
+  /**
+   * Create a new HTML Label Tag based on given name string and attach it to
+   * given parent tag.
+   * @param {string} tagName Type of tag to be created.
+   * @param {string} name InnerHTML of the Label tag.
+   * @param {tag} parentTag Tag to be attached to by the newly created label
+    tag.
+   * @param {string} [forId=undefined] ID to be set as the value of the newly
+    created label tag's for attribute. Needed if Label tag is created for a
+    checkbox.
+   * @return {tag} Newly created Label tag.
+   */
+  function createLabel(tagName, name, parentTag, forId = undefined) {
+    const label = document.createElement(tagName);
+    if (forId) {
+      label.setAttribute('for', forId);
+    }
+    label.innerHTML = name;
+    parentTag.appendChild(label);
+    return label;
+  }
+
+  /**
+   * Create checkboxes for given structure's layers and child structures.
+   * @param {Object} structureData Structure to be extracted and visualized.
+   * @param {String} curStructureId ID of the HTML div element to attach the
+   * new structure's data.
+   * @param {Object} scene object to attach the THREE.js objects
+   * @param {number} thisStructureIndex the index of the structure being
+   processed
+   * generated from structure layer data
+   */
+  function extractSingleStructureData(structureData, curStructureId, scene,
```

Comment:
good start, but i think this function can be refactored down a little bit more. right now it seems like it's doing a lot of the sidebar stuff + rendering the scene. we should decouple that and have explicit pure render functions and other data helper functions

Reply:
separated rendering functions from sidebar creation.